thread/synchronization/thread_synchronization.odin

```diff
@@ -0,0 +1,145 @@
+
+			// First we try to pop the first element off of our queue.
+			// `pop_front_safe` returns the data type of the dynamic array, and a boolean indicating whether it was possible to pop an element or not.
+			// That means we can use `or_break` to break out of our loop if `pop_front_safe` tells us there's nothing left in the queue!
+			item := pop_front_safe(queue) or_break loop
+
+			// Now that we've got the data we need from the queue, we can unlock our mutex to allow for other threads do some work.
+			sync.mutex_unlock(mutex)
+
+			// Now we can do our item processing! Which in this case is just "processing" it for an amount of time, and then printing! :D
+			// Since `processing_time` is a f32, you need to cast `time.Second` to a f32, then back to `time.Duration` to get your fraction of a second.
+			time.sleep(time.Duration(f32(time.Second) * item.processing_time))
+
+			// After we've done our "processing" (sleeping on the job), we can print out some info to the console about our item, and the thread that grabbed it.
+			// `fmt.printfln` and others are thread-safe, so nothing to worry about here.
+			fmt.printfln(
+				"[THREAD %02d] Item %04d processed in %0.2f seconds.",
+				thread_identifier,
+				item.item_tag,
+				item.processing_time,
+			)
+		} else {
+			// Unfortunately, our mutex is locked already! But we're patient, so we can just wait and try again.
+			// We'll sleep for a few milliseconds, and then the loop will continue.
+			time.sleep(time.Millisecond * 5)
+		}
+	}
+}
+
+main :: proc() {
+	// This `RANDOM_SEED` is just a compile-time constant that will seed the default random generator if specified as a non-zero value.
+	// I added this in to allow for predictable, reproducible outputs.
+	//
+	// When it's 0, Odin's random generator is seeded as it normally would be by default.
+	// Otherwise, this `when` clause kicks in at compile-time and will override the default seeding mechanism.
+	//
+	// To specify it yourself, you can just add `--define:RANDOM_SEED=...` to your `odin build/run` command.
+	RANDOM_SEED: u64 : #config(RANDOM_SEED, 0)
+	when RANDOM_SEED > 0 {
+		state := rand.create(RANDOM_SEED)
+		context.random_generator = rand.default_random_generator(&state)
+	}
+
+	// Here we're going to set up a randomized set of data to work off of.
+	// It'll be a dynamic array of `Work_Items`, which essentially just have an ID number and a duration.
+	queue := create_randomized_queue(500)
+
+	// This is a Mutex. If you don't know what that is, it stands for "mutual exclusion lock".
+	// It doesn't actually hold any data, but rather it's used in multi-threaded applications as a way to tell other threads when it's safe to do something.
+	//
+	// A Mutex starts in an UNLOCKED state. At any time, you can LOCK a Mutex using `sync.lock`.
+	// If a Mutex is LOCKED, that means when something else tries to LOCK it, it will fail, since someone else has already LOCKED it.
```

Comment:
It won't fail, unless `try_lock` is used. Typically locking blocks the execution of the current thread until it is unlocked.

```suggestion
	// If a Mutex is LOCKED, that means when another thread tries to LOCK it, it will the execution of the thread will stop, until
	// the mutex is UNLOCKED again. The section of the code between LOCK and UNLOCK operations
	// is called the Critical Section. As such, only one thread at a time can be executing
	// the code inside the critical section.
```